modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpServerTransport.java

```diff
@@ -368,30 +402,52 @@ protected void initChannel(Channel ch) throws Exception {
             if (transport.readTimeoutMillis > 0) {
                 ch.pipeline().addLast("read_timeout", new ReadTimeoutHandler(transport.readTimeoutMillis, TimeUnit.MILLISECONDS));
             }
-            final HttpRequestDecoder decoder = new HttpRequestDecoder(
-                handlingSettings.maxInitialLineLength(),
-                handlingSettings.maxHeaderSize(),
-                handlingSettings.maxChunkSize()
-            );
+            final HttpRequestDecoder decoder;
+            if (headersAuthenticator != null) {
+                decoder = new HttpRequestDecoder(
+                    handlingSettings.maxInitialLineLength(),
+                    handlingSettings.maxHeaderSize(),
+                    handlingSettings.maxChunkSize()
+                ) {
+                    @Override
+                    protected HttpMessage createMessage(String[] initialLine) throws Exception {
+                        DefaultHttpRequest httpRequest = (DefaultHttpRequest) super.createMessage(initialLine);
+                        return headersAuthenticator.wrapNewMessage(httpRequest);
+                    }
+
+                    @Override
+                    protected HttpMessage createInvalidMessage() {
+                        DefaultHttpRequest httpRequest = (DefaultHttpRequest) super.createInvalidMessage();
+                        return headersAuthenticator.wrapNewMessage(httpRequest);
+                    }
+                };
+            } else {
```

Comment:
This is intercepting the netty `HttpMessage` contruction during http request decoding, in order to wrap the default messages with a custom type that in addition carries the authentication context.
The authentication context is carried from the new channel inbound handler, through the netty pipeline, and the ES core code handling HTTP/1.1. pipelining, to the point where the thread context is prepared just before the request is dispatched to the handler.

Reply:
This is the first of two changes to the netty pipeline.